anonip.py

```diff
@@ -374,17 +383,26 @@ def main():
         args.skip_private,
     )
 
-    if args.output:
-        try:
-            with open(args.output, "a") as output_file:
-                for line in anonip.run():
-                    print(line, file=output_file)
-                    output_file.flush()
-        except IOError as err:  # pragma: no cover
-            logger.error(err)
-    else:
-        for line in anonip.run():
-            print(line)
+    input_file = output_file = None
+    try:
+        if args.input:
+            input_file = open(args.input, "r")
```

Comment:
In order to make sure the files are closed properly, we should consider using [ExitStack](https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack) for opening them.

Unfortunately it's only available for python >= 3.3. So we'd have to handle python 2.7 separately. There is [contextlib2](https://github.com/jazzband/contextlib2) but I'd rather not introduce a new dependency.

I've never used this so far, so I was interested in playing around with it. Here is a quick and dirty example that fails on python2.7:

```diff
diff --git a/anonip.py b/anonip.py
index d18e9b2..537c208 100755
--- a/anonip.py
+++ b/anonip.py
@@ -43,6 +43,8 @@ import argparse
 import logging
 import sys
 from io import open
+from contextlib import ExitStack
+
 
 try:
     import ipaddress
@@ -317,9 +319,11 @@ def parse_arguments(args):
     )
     parser.set_defaults(increment=0)
     parser.add_argument("-o", "--output", metavar="FILE", help="file to write to")
+    parser.set_defaults(output=sys.stdout)
     parser.add_argument(
         "--input", metavar="FILE", help="File or FIFO to read from (default: stdin)"
     )
+    parser.set_defaults(input=sys.stdin)
     parser.add_argument(
         "-c",
         "--column",
@@ -383,26 +387,23 @@ def main():
         args.skip_private,
     )
 
-    input_file = output_file = None
+    filenames = {"input": (args.input, "r"), "output": (args.output, "a")}
+
     try:
-        if args.input:
-            input_file = open(args.input, "r")
-        if args.output:
-            output_file = open(args.output, "a")
-        else:
-            output_file = sys.stdout
-        for line in anonip.run(input_file):
-            print(line, file=output_file)
-            # TODO: when dropping support for Python <= 3.3, move the
-            # flush into the print()
-            output_file.flush()
+        with ExitStack() as stack:
+            files = {
+                key: stack.enter_context(open(fname[0], fname[1]))
+                if isinstance(fname[0], str)
+                else fname[0]
+                for key, fname in filenames.items()
+            }
+            for line in anonip.run(files["input"]):
+                print(line, file=files["output"])
+                # TODO: when dropping support for Python <= 3.3, move the
+                # flush into the print()
+                files["output"].flush()
     except IOError as err:  # pragma: no cover
         logger.error(err)
-    finally:
-        if args.input and input_file:
-            input_file.close()
-        if args.output and output_file:
-            output_file.close()
 
 
 if __name__ == "__main__":  # pragma: no cover
```

Reply:
> In order to make sure the files are closed properly, 

This is exactly what the `finally` part in the current code does.

> we should consider using ExitStack for opening them.

ExitStack would indeed be a nice solution. But since the code should be Python 2.7 compatilbe and adding another dependency this is not an option, I suggest to have to code as proposed by me.